src/form_beam.cpp

```diff
@@ -248,128 +256,101 @@ __global__ void vmBeamform_kernel( gpuDoubleComplex *Jv_Q,
                                  int npol,
                                  int nstokes )
 {
-    // Translate GPU block/thread numbers into meaningful names
-    int c    = blockIdx.x;  /* The (c)hannel number */
-    int nc   = gridDim.x;   /* The (n)umber of (c)hannels (=128) */
-    int s    = blockIdx.y;  /* The (s)ample number */
-    int ns   = gridDim.y;   /* The (n)umber of (s)amples (in a chunk)*/
 
-    int ant  = threadIdx.x; /* The (ant)enna number */
-    int nant = blockDim.x;  /* The (n)_umber of (ant)ennas */
-
-    // Organise dynamically allocated shared arrays (see tag 11NSTATION for kernel call)
-    extern __shared__ double arrays[];
-    // These SHOULD be *aligned* on integer numbers of 4-byte blocks.
-
-    // NOTE: Because these are complex-doubles, each takes up 2*sizeof(double), hence the stride of 2 here.
-    /* Given that we need to ensure access alignment on the 4-byte boundaries (CUDA requirement), we have
-       to make sure that the array access corresponds to an integer number of sizeof(double), which
-       means we need to use even indexes to access memory. This ensures for all `nant` that we access
-       within the memory boundaries. This StackOverflow post helped us figure this out:
-         
-            https://stackoverflow.com/questions/70765553/cuda-shared-memory-alignement-in-documentation
-
-       (Previously, the indexes were: 1*nant, 3*nant, etc.)
+    const unsigned int warp_id {threadIdx.x / warpSize};
```

Comment:
I'm not familiar with this syntax. I realise it's maybe just inexperience, but a comment explaining it (or noting the syntax definition for someone to look up) could be helpful for future development? Bearing in mind, most people who will maintain this are not necessarily software engineers/GPU experts.

Reply:
> I'm not familiar with this syntax.

It is a C++ language feature called [initialiser list]( https://en.cppreference.com/w/cpp/language/list_initialization). Given that the code is written with C syntax I should better stick with a classic assignment, that is

```
const unsigned int warp_id = threadIdx.x / warpSize;
```

> Is there any sense in promoting things to doubles during the beamforming stage and then demoting back to floats on output?

Not really. Between the results of the current implementation and the one proposed in this PR, the latter has the most accurate result. This is because in parallel reduction, assuming the starting values are all in the same order of magnitude, partial sums are still in similar order of magnitude (each partial sum is the sum of the same number of values). In the sequential implementation, the value in the accumulation variable keeps growing order of magnitude larger than the values being added. [And it is known](https://stackoverflow.com/questions/22186589/why-does-adding-a-small-float-to-a-large-float-just-drop-the-small-one) that when adding a very small value to a larger one, that contribution can get lost.